### PR TITLE
[ECO-4645] Move `React.createContext` outside of components in react-hooks

### LINKED ISSUE #1638: Move `React.createContext` outside of components
Currently [we call](https://github.com/ably/ably-js/blob/64caf8e52152a1ed7b0c1ca32689e2e45369dca8/src/platform/react-hooks/src/AblyProvider.tsx#L40) `React.createContext` during render in `AblyProvider`, so we can have arbitrary amount of contexts accessible by separate `id` value. It is considered best practice to create a react context outside of any component to avoid context re-creation on every render.

We can move `React.createContext` outside of `AblyProvider` if we change value stored in the context from:
```javascript
type AblyContextProps = {
  client: Ably.RealtimeClient;
  _channelNameToInstance: Record<string, Ably.RealtimeChannel>;
}
```

to:
```javascript
type AblyContextProps = Record<string, {
  client: Ably.RealtimeClient;
  _channelNameToInstance: Record<string, Ably.RealtimeChannel>;
}>
```

with keys in Record being provider ids (same `id` field used before), so it will store values like this:
```javascript
{
  providerId1: {
    client,
    _channelNameToInstance: {
      'channel-name-1': {
        // ... options
      }
    },
  },
  providerId2: {
    client,
    _channelNameToInstance: {
      'channel-name-2': {
        // ... options
      }
    },
  },
  ...
}
```

And then `AblyProvider` can have something like this to update context value:
```javascript
export const AblyContext = React.createContext<AblyContextProps>({});

export const AblyProvider = ({ client, children, id = 'default' }) => {
  const context = React.useContext(GlobalAblyContext);

  const value: AblyContextProps = useMemo(() => {
    return { ...context, [id]: { client, _channelNameToInstance: {} } };
  }, [context, client, id]);

  return <GlobalAblyContext.Provider value={value}>{children}</GlobalAblyContext.Provider>;
};

```

┆Issue is synchronized with this [Jira Task](https://ably.atlassian.net/browse/ECO-4645) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by VeskeR**
Resolves #1638

### CODE REVIEW

File: src/platform/react-hooks/src/AblyProvider.tsx [hunk 1]
```diff
@@ -1,50 +1,33 @@
...
+// An object which stores all provider options indexed by provider id,
+// which is used to get options set by specific `AblyProvider` after calling `React.useContext`.
+export type AblyContextType = Record<string, AblyContextProviderProps>;
 
-const ctxMap: ContextMap = typeof globalThis !== 'undefined' ? (globalThis[contextKey] = {}) : {};
-
-export function getContext(ctxId = 'default'): AblyContextType {
-  return ctxMap[ctxId];
-}
+export const AblyContext = React.createContext<AblyContextType>({});
```
[the remark this answers is not in the document]
Comment (VeskeR) [reply src/platform/react-hooks/src/AblyProvider.tsx]: Restored previous logic with comments describing the chosen approach of storing context in the global object.

Also moved context related code to new `AblyContext` - no point to have it in `AblyProvider`.

**Review by ttypic — COMMENTED** [review]
I like it, there are some conflicts, but everything looks good to me

**Comment by VeskeR** [comment]
Changed this PR to be based on https://github.com/ably/ably-js/pull/1676 - that one has higher priority and I would like to merge it first.

Resolved conflicts - no functional changes to the implementation, so we can merge as soon as https://github.com/ably/ably-js/pull/1676 is merged.

**Review by ttypic — APPROVED** [review]
LGTM

**Pull request merged**